torch/_inductor/fx_passes/decompose_mm.py

```diff
@@ -0,0 +1,551 @@
+
+    post_mm_ops = []
+    reshape_ops_count = 0
+    view3d_node = None
+
+    for node in reversed(intermediate_nodes):
+        if node.target in (aten.view.default, aten.reshape.default):
+            in_shape = node.args[0].meta["val"].shape
+            out_shape = node.args[1]
+
+            # Support only 2D -> 3D view with matching last dimension
+            if not (
+                len(in_shape) == 2
+                and len(out_shape) == 3
+                and in_shape[-1] == out_shape[-1]
+            ):
+                return None, None
+
+            # Support only one 2D -> 3D reshape
+            reshape_ops_count += 1
+            if reshape_ops_count > 1:
+                return None, None
+
+            post_mm_ops.append(_ViewOp(node, chunk_dim=0))
+            view3d_node = node
+            continue
+
+        if not _is_pointwise(node):
+            return None, None
+
+        post_mm_ops.append(_PostMMOp(node))
+
+    return post_mm_ops, view3d_node
+
+
+def split_mm(
+    gm: torch.fx.GraphModule, min_size_after_split: int, num_chunks: list[int]
+):
+    g = gm.graph
+    g.owning_module
+
+    for n in g.nodes:
+        if not (n.op == "call_function" and n.target == torch.ops.aten.mm.default):
+            continue
+
+        mm_n = n
+        arg_a = mm_n.args[0]
+        arg_b = mm_n.args[1]
+
+        a_t = arg_a.meta["val"]
+
+        size_before_split = 1
+        for s in a_t.shape[:-1]:
+            size_before_split *= _size_hint(s)
+
+        n_split = _select_optimal_chunk_count(
+            size_before_split, num_chunks, min_size_after_split
+        )
+
+        if n_split is None:
+            return
+
+        arg_a_t = arg_a.meta["val"]
+        arg_b_t = arg_b.meta["val"]
+        mm_out_t = mm_n.meta["val"]
+
+        # Decompose only to contiguous chunks
+        if not (_is_contiguous(arg_a_t) and _is_contiguous(mm_out_t)):
+            continue
+
+        trace_args = (arg_a_t, arg_b_t, n_split, mm_out_t)
+        _insert_fn_trace_before_node(
+            g,
+            _fn,
+            trace_args,
+            mm_n,  # insert before
+            [arg_a, arg_b],
+            [mm_n],
+        )
+
+        g.erase_node(mm_n)
+
+
+@dataclasses.dataclass
+class _ReduceScatterMatch:
+    match: Match
+    input_node: torch.fx.Node
+    reduce_scatter_node: torch.fx.Node
+    wait_tensor_node: torch.fx.Node
+    reduce_op: str
+    scatter_dim: int
+    group_size: int
+    group_name: str
+
+
+def _split_mm_rs(
+    gm: torch.fx.GraphModule,
+    num_chunks: list[int],
+    min_size_after_split: int = 2048,
+):
+    """
+    Splits matmul->reduce_scatter into chunks,
+    number of chunks will be the greatest value from num_chunks,
+    with condition that matmuls dimension after split is larger than min_size_after_split.
+
+    The main motivation usecase is overlap scheduling pass. Splitting large matmuls helps
+    more fine grained overlapping.
```

Comment:
out of curiosity:

(1) do we do anything to make sure that the chunked matmul is not "too small" (small enough to not have good perf)?

(2) if there is a large matmul that we actually would have been able to schedule/overlap, do we figure this out and avoid chunking it?